### ISSUE TITLE
Exoplayer OOM

### ISSUE THREAD

**Issue by IriaSomobu** [TeamNewPipe/NewPipe#8282]
### Checklist

- [X] I am able to reproduce the bug with the [latest version](https://github.com/TeamNewPipe/NewPipe/releases/latest).
- [X] I made sure that there are *no existing issues* - [open](https://github.com/TeamNewPipe/NewPipe/issues) or [closed](https://github.com/TeamNewPipe/NewPipe/issues?q=is%3Aissue+is%3Aclosed) - which I could contribute my information to.
- [X] I have taken the time to fill in all the required details. I understand that the bug report will be dismissed otherwise.
- [X] This issue contains only one bug.
- [X] I have read and understood the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/dev/.github/CONTRIBUTING.md).

### Affected version

0.22.2 commit 53bf342

### Steps to reproduce the bug

1. Open NewPipe on rather long YT video (like [this](https://www.youtube.com/watch?v=LzUT1X10Ihk));
2. Wait for about 2 minutes;
3. OOM is here

Sometimes you can avoid OOM by switching resolution before initial video loading is done.

### Expected behavior

Video should be played w/o crashes

### Actual behavior

 - Sometimes NewPipe just stops playing video with error notification;
 - Sometimes NewPipe crashes with ANR.

### Logs

```
com.google.android.exoplayer2.ExoPlaybackException: Source error
   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:641)
   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:617)
   at android.os.Handler.dispatchMessage(Handler.java:98)
   at android.os.Looper.loop(Looper.java:154)
   at android.os.HandlerThread.run(HandlerThread.java:61)
Caused by: com.google.android.exoplayer2.upstream.Loader$UnexpectedLoaderException: Unexpected OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack trace available
   at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:441)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
   at java.lang.Thread.run(Thread.java:761)
Caused by: java.lang.OutOfMemoryError: OutOfMemoryError thrown while trying to throw OutOfMemoryError; no stack trace available
```

### Affected Android/Custom ROM version

RR-N-v5.8.5-20171005-ivy-Unofficial

### Affected device model

Sony E6533

### Additional information

I've reported that I can reproduce ExoPlayer OOM in #2257 and was told to open a new issue. So here I am.

**Comment by TiA4f8R** [comment]
Yes, I can reproduce this issue too on a "200H video" too. This issue happens because we are right now using progressive media sources. Because ExoPlayer doesn't know the file structure (format, metadata, sidx boxes, etc), it needs to download first a part of the file, which is so really big for long videos and this should be the cause of the OOM error you get.

What happens when you use the APK #8153, which uses DASH, like YouTube clients, and so downloads smaller video and audio chunks? I tested the video with my changes and playback works fine.

**Comment by litetex** [comment]
> RR-N-v5.8.5-20171005-ivy-Unofficial

What operating system is this? Doesn't look like normal Android...

**Comment by oddtoddy** [comment]
I think that's Remix OS

**Comment by IriaSomobu** [comment]
> What happens when you use the APK https://github.com/TeamNewPipe/NewPipe/pull/8153

@TiA4f8R I can confirm that your PR makes initial loading faster and fixes OOM thrown before video started playing. (I've tested [this](https://www.youtube.com/watch?v=xT2ZOZhn5yQ) vid)

But I've seen another OOM pattern:
1. Video starts playing;
2. After 1-5 minutes there are a few consequent freezes, but video continues to play;
3. App crashes with ANR.

I cannot find a way to reproduce this scenario right now, but I'll test your PR against this case as soon as I stumble on such vid.

> What operating system is this?

@litetex, its [Resurrection Remix](https://resurrectionremix.com/). In my case, Android 7.1.2 on Linux 3.10.84.

**Comment by litetex** [comment]
I think the problem is likely caused by a OS problem which can't be fixed by NewPipe:
1. You are using a phone released in 2015. As far as I can see it has 3GB of RAM so this should be sufficient to run NewPipe (at least when nothing else is open and consuming massive amounts or RAM).
When I view the above video NewPipe needs a maximum of 200MB, my system/phone in total 2GB out of 4GB and everything works fine (although it's a bit slower than usual)
2. That phone is normally shipped with Android 5 or 6 (according to https://www.gsmarena.com/sony_xperia_z3+_dual-7257.php). However you are using a custom Android 7. I'm not sure if the phone behaves correctly with this OS.

I would recommend the following steps:
1. Did you restart your phone?
2. Check your settings / memory usage, is something else consuming all the RAM?
3. Try update the custom ROM if possible
4. If updating doesn't fix the problem try use the manufacturers official OS

**Comment by IriaSomobu** [comment]
Yep, looks like the problem appears in very specific conditions: I'm unable to trigger it on latest dev version while debugging. Upd: now I'm unable to trigger it on latest f-droid release too.

<details>
  <summary>As for RAM consuming, normally I see something like this</summary>

| Figure A | Figure B |
| --- | --- |
| ![Screenshot_20220424-190014](https://user-images.githubusercontent.com/17745136/165044398-585f2a3e-6c44-4886-ac00-40cac5068517.png) | ![Screenshot_20220424-190021](https://user-images.githubusercontent.com/17745136/165044403-ae6261ce-bad8-4af2-8c92-153ddfa85a18.png) |

</details>

I'm not ready to tinker with custom or official ROMs right now, sorry.

To sum up: I'm unable to reproduce problem right now, but I think it might occur again in the future. When it happens, I'll collect debug info and try what you suggested.

For now I think its better to close this issue.
